src/types.rs

```diff
@@ -522,51 +534,135 @@ pub enum DateError {
     /// The month is out of bounds.
     #[error("month not in interval 1-12")]
     MonthOutOfBounds,
+    /// The offset is out of bounds.
+    #[error("offset not in interval -24:00 -- 24:00")]
+    OffsetOutOfBounds,
+}
+
+impl Display for Date {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let Date {
+            year,
+            month,
+            day,
+            hour,
+            minute,
+            second,
+            nanosecond,
+            offset,
+        } = self;
+        write!(f, "{year:04}")?;
+
+        if let Some(month) = month {
+            let month = month + 1;
+            write!(f, "-{month:02}")?;
+        } else {
+            return Ok(());
+        }
+
+        if let Some(day) = day {
+            let day = day + 1;
+            write!(f, "-{day:02}")?;
+        } else {
+            return Ok(());
+        }
+
+        if let Some(hour) = hour {
+            write!(f, "T{hour:02}")?;
+        } else {
+            return Ok(());
+        }
+
+        if let Some(minute) = minute {
+            write!(f, ":{minute:02}")?;
+            if let Some(second) = second {
+                write!(f, ":{second:02}")?;
+                if let Some(nanosecond) = nanosecond {
+                    write!(f, ".{nanosecond:09}")?;
+                }
+            }
+        }
+
+        if let Some(offset) = offset {
+            write!(f, "{offset}")?;
+        }
+
+        Ok(())
+    }
 }
 
 impl FromStr for Date {
     type Err = DateError;
 
     /// Parse a date from a string.
     fn from_str(source: &str) -> Result<Self, Self::Err> {
+        use chrono::format::{self, Fixed, Item, ParseErrorKind, Parsed};
         let mut source = source.to_string();
         source.retain(|f| !f.is_whitespace());
 
-        let full_date = source.parse::<NaiveDate>();
+        let mut parsed = Parsed::new();
+        match format::parse(
+            &mut parsed,
+            &source,
+            std::iter::once(Item::Fixed(Fixed::RFC3339)),
+        ) {
+            Ok(()) => (),
+            Err(error) if error.kind() == ParseErrorKind::TooShort => (),
+            Err(_) => todo!(),
```

Comment:
Short dates occur when month or day are missing. Since these are not required, failing here would limit the parser to full datetimes with offsets.

All other errors from chrono are now correctly propagated.